src/basicfuns.jl

```diff
@@ -190,7 +190,13 @@ function logsumexp(X)
     isempty(X) && return log(sum(X))
     reduce(logaddexp, X)
 end
-function logsumexp(X::AbstractArray{T}) where {T<:Real}
+function logsumexp(X::AbstractArray{T}; dims=nothing) where {T<:Real}
+    dims isa Nothing && return _logsumexp(X)
+    isempty(X) && return log(zero(T))
+    u = maximum(X; dims=dims)
+    return log.(sum(exp.(X .- u); dims=dims)) .+ u
```

Comment:
Maybe reuse the array created by `sum(exp.(X .- u); dims=dims)` to avoid a temporary.